# Feature_Engineering 목표

1. users, vehicles 데이터를 AccidentId 기준으로 집계하여 1:1 구조로 변환
2. 집계된 피처를 accidents, places와 병합하여 사고 단위의 학습 데이터 구성
4. 범주형 변수 인코딩 및 최종 학습모델 생성

In [1]:
# 필수 라이브러리
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib as mpl

# 한글 폰트 설정 (Windows - 맑은 고딕)
mpl.rc('font', family='Malgun Gothic')
mpl.rcParams['axes.unicode_minus'] = False  # 마이너스 기호 깨짐 방지

In [2]:
# 데이터 불러오기
accidents_train = pd.read_csv('./data/Cleaning/accidents_train.csv')
accidents_test = pd.read_csv('./data/Cleaning/accidents_test.csv')

places_train = pd.read_csv('./data/Cleaning/places_train.csv')
places_test = pd.read_csv('./data/Cleaning/places_test.csv')

users_train = pd.read_csv('./data/Cleaning/users_train.csv')
users_test = pd.read_csv('./data/Cleaning/users_test.csv')

vehicles_train = pd.read_csv('./data/Cleaning/vehicles_train.csv')
vehicles_test = pd.read_csv('./data/Cleaning/vehicles_test.csv')

In [3]:
users_train.info()

<class 'pandas.DataFrame'>
RangeIndex: 86992 entries, 0 to 86991
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   AccidentId          86992 non-null  int64  
 1   VehicleId           86992 non-null  str    
 2   Seat                78586 non-null  float64
 3   Category            86992 non-null  str    
 4   Gender              86992 non-null  str    
 5   TripReason          86992 non-null  str    
 6   SafetyDevice        86992 non-null  str    
 7   SafetyDeviceUsed    86992 non-null  str    
 8   PedestrianLocation  86992 non-null  str    
 9   PedestrianAction    86992 non-null  str    
 10  PedestrianCompany   86992 non-null  str    
 11  BirthYear           83996 non-null  float64
 12  is_pedestrian       86992 non-null  int64  
 13  target_group_size   1419 non-null   float64
 14  RoadUserType        86992 non-null  str    
 15  Age                 86992 non-null  str    
 16  Age_missing_fla

In [4]:
users_train.head(20)

,AccidentId,VehicleId,Seat,Category,Gender,TripReason,SafetyDevice,SafetyDeviceUsed,PedestrianLocation,PedestrianAction,PedestrianCompany,BirthYear,is_pedestrian,target_group_size,RoadUserType,Age,Age_missing_flag,AgeGroup
0,201800000002,A01,1.0,PassengerCar,Male,Unknown,SeatBelt,Yes,Not_Applicable,Not_Applicable,Unknown,1947.0,0,NaN,Driver,71,0,70+
1,201800000002,A01,NaN,PassengerCar,Male,Unknown,Helmet,Missing,OnLane<=OnSidewalk0mCrossing,Crossing,Alone,1965.0,1,NaN,Pedestrian,53,0,le_60
2,201800000003,A01,1.0,TwoWheeler,Male,Leisure,Helmet,Yes,Not_Applicable,Not_Applicable,Unknown,1937.0,0,NaN,Driver,81,0,70+
3,201800000003,B01,NaN,PassengerCar,Unknown,Unknown,Missing,Missing,Not_Applicable,Not_Applicable,Unknown,NaN,0,1.0,Driver,Unknown,1,Unknown
4,201800000003,C01,1.0,PassengerCar,Male,Unknown,ChildrenDevice,Missing,Not_Applicable,Not_Applicable,Unknown,NaN,0,NaN,Driver,Unknown,1,Unknown
5,201800000004,A01,1.0,PassengerCar,Male,Leisure,SeatBelt,Yes,Not_Applicable,Not_Applicable,Unknown,1982.0,0,NaN,Driver,36,0,le_40
6,201800000004,B01,1.0,Bicycle,Male,Leisure,Helmet,Missing,Not_Applicable,Not_Applicable,Unknown,2011.0,0,NaN,Driver,7,0,le_10
7,201800000008,A01,1.0,PassengerCar,Male,Other,SeatBelt,Yes,Not_Applicable,Not_Applicable,Unknown,1987.0,0,NaN,Driver,31,0,le_40
8,201800000008,A01,NaN,PassengerCar,Male,Leisure,Helmet,Missing,OnCrossingWithLigths,Crossing,Alone,1944.0,1,NaN,Pedestrian,74,0,70+
9,201800000011,A01,1.0,PassengerCar,Female,Shopping,SeatBelt,Yes,Not_Applicable,Not_Applicable,Unknown,1953.0,0,NaN,Driver,65,0,le_70


In [5]:
vehicles_train.info()

<class 'pandas.DataFrame'>
RangeIndex: 63500 entries, 0 to 63499
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   AccidentId        63500 non-null  int64
 1   Direction         63500 non-null  str  
 2   FixedObstacle     63500 non-null  str  
 3   MobileObstacle    63500 non-null  str  
 4   ImpactPoint       63500 non-null  str  
 5   Maneuver_grouped  63500 non-null  str  
dtypes: int64(1), str(5)
memory usage: 2.9 MB


In [6]:
vehicles_train.head(20)

,AccidentId,Direction,FixedObstacle,MobileObstacle,ImpactPoint,Maneuver_grouped
0,201800000002,Unknown,Not_Applicable,Pedestrian,Unknown,Straight
1,201800000003,Unknown,StationaryVehicle,Vehicle,Front,Straight
2,201800000003,Unknown,Not_Applicable,Vehicle,LeftSide,Turn
3,201800000003,Unknown,Not_Applicable,Not_Applicable,RightSide,StoppedOrParked
4,201800000004,Unknown,Not_Applicable,Other,RightFront,ModerateRisk
5,201800000004,Unknown,Not_Applicable,Vehicle,LeftSide,Unknown
6,201800000008,Unknown,Not_Applicable,Pedestrian,RightFront,Straight
7,201800000011,Unknown,Not_Applicable,Other,LeftFront,Turn
8,201800000011,Unknown,Not_Applicable,Vehicle,Front,Unknown
9,201800000012,Increasing,Post,Not_Applicable,Unknown,Unknown


# Uers 결합테이블 생성하기

In [7]:
# users_train → AccidentId 단위 집계

# 1. 전체 사람 수
people_count = (
    users_train
    .groupby('AccidentId')
    .size()
    .rename('PeopleCount')
)

# 2. 운전자 / 동승자 / 보행자 수
role_count = pd.crosstab(
    users_train['AccidentId'],
    users_train['RoadUserType']
)

# 실제 존재하는 컬럼만 사용하기 위해 확인
role_count = role_count.rename(columns={
    'Driver': 'DriverCount',
    'Passenger': 'PassengerCount',
    'Pedestrian': 'PedestrianCount'
})

# 필요한 컬럼이 없는 경우 0으로 생성
for col in ['DriverCount', 'PassengerCount', 'PedestrianCount']:
    if col not in role_count.columns:
        role_count[col] = 0

role_count = role_count[
    ['DriverCount', 'PassengerCount', 'PedestrianCount']
]


# 3. 사고에 참여한 차량 수
vehicle_count = (
    users_train
    .groupby('AccidentId')['VehicleId']
    .nunique()
    .rename('VehicleCount')
)


# 4. Category 원-핫 → 사고별 차량 유형 수
category_ohe = pd.get_dummies(
    users_train['Category'],
    prefix='Category',
    dtype=int
)

category_ohe['AccidentId'] = users_train['AccidentId'].values

category_count = (
    category_ohe
    .groupby('AccidentId')
    .sum()
)


# 5. 성별 인원
gender_count = pd.crosstab(
    users_train['AccidentId'],
    users_train['Gender']
)

gender_count = gender_count.rename(columns={
    'Male': 'MaleCount',
    'Female': 'FemaleCount'
})

for col in ['MaleCount', 'FemaleCount']:
    if col not in gender_count.columns:
        gender_count[col] = 0

gender_count = gender_count[
    ['MaleCount', 'FemaleCount']
]


# 6. TripReason 원-핫 → 사고별 목적 유형 수
trip_ohe = pd.get_dummies(
    users_train['TripReason'],
    prefix='TripReason',
    dtype=int
)

trip_ohe['AccidentId'] = users_train['AccidentId'].values

trip_count = (
    trip_ohe
    .groupby('AccidentId')
    .sum()
)


# 7. SafetyDevice 원-핫
device_ohe = pd.get_dummies(
    users_train['SafetyDevice'],
    prefix='SafetyDevice',
    dtype=int
)

device_ohe['AccidentId'] = users_train['AccidentId'].values

device_count = (
    device_ohe
    .groupby('AccidentId')
    .sum()
)


# 8. SafetyDeviceUsed 원-핫
device_used_ohe = pd.get_dummies(
    users_train['SafetyDeviceUsed'],
    prefix='SafetyDeviceUsed',
    dtype=int
)

device_used_ohe['AccidentId'] = users_train['AccidentId'].values

device_used_count = (
    device_used_ohe
    .groupby('AccidentId')
    .sum()
)


# 9. is_pedestrian
# 사고에 보행자가 한 명이라도 있으면 1
pedestrian_flag = (
    users_train
    .groupby('AccidentId')['is_pedestrian']
    .max()
    .rename('is_pedestrian')
)


# 10. AgeGroup 원-핫 → 사고별 연령대 구성
age_ohe = pd.get_dummies(
    users_train['AgeGroup'],
    prefix='AgeGroup',
    dtype=int
)

age_ohe['AccidentId'] = users_train['AccidentId'].values

age_count = (
    age_ohe
    .groupby('AccidentId')
    .sum()
)


# 11. age_missing_count
# Age_missing_flag → 사고별 결측 인원 수
age_missing_count = (
    users_train
    .groupby('AccidentId')['Age_missing_flag']
    .sum()
    .rename('age_missing_count')
)

In [8]:
# 모든 데이터 결합

users_train_new = pd.concat([
    people_count,
    role_count,
    vehicle_count,
    category_count,
    gender_count,
    trip_count,
    device_count,
    device_used_count,
    pedestrian_flag,
    age_count,
    age_missing_count
], axis=1)


# 결측값 → 0
users_train_new = users_train_new.fillna(0)


# AccidentId를 일반 컬럼으로 변환
users_train_new = users_train_new.reset_index()

In [9]:
# users_test 생성 AccidentId 단위 집계

# 1. 전체 사람 수
people_count = (
    users_test
    .groupby('AccidentId')
    .size()
    .rename('PeopleCount')
)


# 2. 운전자 / 동승자 / 보행자 수
role_count = pd.crosstab(
    users_test['AccidentId'],
    users_test['RoadUserType']
)

role_count = role_count.rename(columns={
    'Driver': 'DriverCount',
    'Passenger': 'PassengerCount',
    'Pedestrian': 'PedestrianCount'
})

for col in ['DriverCount', 'PassengerCount', 'PedestrianCount']:
    if col not in role_count.columns:
        role_count[col] = 0

role_count = role_count[
    ['DriverCount', 'PassengerCount', 'PedestrianCount']
]


# 3. 차량 수
vehicle_count = (
    users_test
    .groupby('AccidentId')['VehicleId']
    .nunique()
    .rename('VehicleCount')
)


# 4. Category
category_ohe = pd.get_dummies(
    users_test['Category'],
    prefix='Category',
    dtype=int
)

category_ohe['AccidentId'] = users_test['AccidentId'].values

category_count = (
    category_ohe
    .groupby('AccidentId')
    .sum()
)


# 5. 성별
gender_count = pd.crosstab(
    users_test['AccidentId'],
    users_test['Gender']
)

gender_count = gender_count.rename(columns={
    'Male': 'MaleCount',
    'Female': 'FemaleCount'
})

for col in ['MaleCount', 'FemaleCount']:
    if col not in gender_count.columns:
        gender_count[col] = 0

gender_count = gender_count[
    ['MaleCount', 'FemaleCount']
]


# 6. TripReason
trip_ohe = pd.get_dummies(
    users_test['TripReason'],
    prefix='TripReason',
    dtype=int
)

trip_ohe['AccidentId'] = users_test['AccidentId'].values

trip_count = (
    trip_ohe
    .groupby('AccidentId')
    .sum()
)


# 7. SafetyDevice
device_ohe = pd.get_dummies(
    users_test['SafetyDevice'],
    prefix='SafetyDevice',
    dtype=int
)

device_ohe['AccidentId'] = users_test['AccidentId'].values

device_count = (
    device_ohe
    .groupby('AccidentId')
    .sum()
)


# 8. SafetyDeviceUsed
device_used_ohe = pd.get_dummies(
    users_test['SafetyDeviceUsed'],
    prefix='SafetyDeviceUsed',
    dtype=int
)

device_used_ohe['AccidentId'] = users_test['AccidentId'].values

device_used_count = (
    device_used_ohe
    .groupby('AccidentId')
    .sum()
)


# 9. is_pedestrian
pedestrian_flag = (
    users_test
    .groupby('AccidentId')['is_pedestrian']
    .max()
    .rename('is_pedestrian')
)


# 10. AgeGroup
age_ohe = pd.get_dummies(
    users_test['AgeGroup'],
    prefix='AgeGroup',
    dtype=int
)

age_ohe['AccidentId'] = users_test['AccidentId'].values

age_count = (
    age_ohe
    .groupby('AccidentId')
    .sum()
)


# 11. age_missing_count
age_missing_count = (
    users_test
    .groupby('AccidentId')['Age_missing_flag']
    .max()
    .rename('age_missing_count')
)

# 모든 데이터 결합
users_test_new = pd.concat([
    people_count,
    role_count,
    vehicle_count,
    category_count,
    gender_count,
    trip_count,
    device_count,
    device_used_count,
    pedestrian_flag,
    age_count,
    age_missing_count
], axis=1)


users_test_new = users_test_new.fillna(0)

users_test_new = users_test_new.reset_index()

In [10]:
# train과 test의 컬럼을 동일하게 맞춤
users_train_new, users_test_new = users_train_new.align(users_test_new,
                                                        join='outer',
                                                        axis=1,
                                                        fill_value=0)

In [11]:
print(users_train_new.shape)
print(users_test_new.shape)

print(users_train_new.columns.equals(users_test_new.columns))

(38257, 45)
(9565, 45)
True


# Vehicles 결합 데이터 생성

In [13]:
# vehicles_train → AccidentId 단위 집계

vehicle_features = [
    'Direction',
    'FixedObstacle',
    'MobileObstacle',
    'ImpactPoint',
    'Maneuver_grouped'
]

vehicle_feature_list = []

for col in vehicle_features:

    ohe = pd.get_dummies(
        vehicles_train[col],
        prefix=col,
        dtype=int
    )

    ohe['AccidentId'] = vehicles_train['AccidentId'].values

    feature_count = (
        ohe
        .groupby('AccidentId')
        .sum()
    )

    vehicle_feature_list.append(feature_count)


vehicles_train_new = pd.concat(vehicle_feature_list,axis=1)
vehicles_train_new = vehicles_train_new.fillna(0)
vehicles_train_new = vehicles_train_new.reset_index()

In [14]:
# vehicles_test → AccidentId 단위 집계

vehicle_features = [
    'Direction',
    'FixedObstacle',
    'MobileObstacle',
    'ImpactPoint',
    'Maneuver_grouped'
]

vehicle_feature_list = []

for col in vehicle_features:

    ohe = pd.get_dummies(
        vehicles_test[col],
        prefix=col,
        dtype=int
    )

    ohe['AccidentId'] = vehicles_test['AccidentId'].values

    feature_count = (
        ohe
        .groupby('AccidentId')
        .sum()
    )

    vehicle_feature_list.append(feature_count)


vehicles_test_new = pd.concat(vehicle_feature_list,axis=1)
vehicles_test_new = vehicles_test_new.fillna(0)
vehicles_test_new = vehicles_test_new.reset_index()

In [15]:
# train / test 컬럼 통일
vehicles_train_new, vehicles_test_new = vehicles_train_new.align(vehicles_test_new,
                                                                 join='outer',
                                                                 axis=1,
                                                                 fill_value=0)

In [16]:
print(vehicles_train_new.shape)
print(vehicles_test_new.shape)

print(vehicles_train_new.columns.equals(vehicles_test_new.columns))

(38257, 48)
(9565, 48)
True


# 데이터셋 Merge

### accidents + places

In [17]:
# merge
train = accidents_train.merge(places_train,
                              on='AccidentId',
                              how='left')

test = accidents_test.merge(places_test,
                            on='AccidentId',
                            how='left')

### +users

In [18]:
train = train.merge(users_train_new,
                    on='AccidentId',
                    how='left')

test = test.merge(users_test_new,
                  on='AccidentId',
                  how='left')

### +vehicles

In [19]:
train = train.merge(vehicles_train_new,
                    on='AccidentId',
                    how='left')

test = test.merge(vehicles_test_new,
                  on='AccidentId',
                  how='left')

In [20]:
print(f'최종 train shape :{train.shape}')
print(f'최종 test shape :{test.shape}')

최종 train shape :(38256, 113)
최종 test shape :(9565, 113)


In [21]:
train.head()

,AccidentId,Gravity,Light,InAgglomeration,IntersectionType,Weather,CollisionType,Month,DayOfWeek,HourOfDay,...,Maneuver_grouped_BusLane,Maneuver_grouped_HighRisk,Maneuver_grouped_LaneChange,Maneuver_grouped_ModerateRisk,Maneuver_grouped_Overtake,Maneuver_grouped_StoppedOrParked,Maneuver_grouped_Straight,Maneuver_grouped_SwerveTurn,Maneuver_grouped_Turn,Maneuver_grouped_Unknown
0,201800023440,NonLethal,TwilightOrDawn,Yes,NoIntersection,Other,Other,2,Friday,7,...,0,0,0,0,0,0,1,0,0,0
1,201800041166,NonLethal,Daylight,Yes,NoIntersection,Normal,Other,10,Wednesday,8,...,0,0,0,0,0,1,0,0,0,0
2,201800040998,NonLethal,NightNoStreetLight,Yes,NoIntersection,Normal,2Vehicles-BehindVehicles-Frontal,12,Thursday,3,...,0,0,0,0,0,0,1,1,0,0
3,201800030984,NonLethal,NightStreelightsOn,Yes,NoIntersection,Normal,2Vehicles-Side,2,Saturday,21,...,0,0,0,0,0,0,1,0,1,0
4,201800036021,NonLethal,Daylight,Yes,NoIntersection,Normal,3+Vehicles-Chain,3,Thursday,13,...,0,0,0,0,0,1,1,0,0,0


In [22]:
test.head()

,AccidentId,Gravity,Light,InAgglomeration,IntersectionType,Weather,CollisionType,Month,DayOfWeek,HourOfDay,...,Maneuver_grouped_BusLane,Maneuver_grouped_HighRisk,Maneuver_grouped_LaneChange,Maneuver_grouped_ModerateRisk,Maneuver_grouped_Overtake,Maneuver_grouped_StoppedOrParked,Maneuver_grouped_Straight,Maneuver_grouped_SwerveTurn,Maneuver_grouped_Turn,Maneuver_grouped_Unknown
0,201800008483,NonLethal,Daylight,Yes,NoIntersection,Normal,Other,8,Thursday,8,...,0,0,0,0,0,0,1,0,0,0
1,201800046518,NonLethal,Daylight,No,NoIntersection,LightRain,2Vehicles-Behind,4,Thursday,17,...,0,0,0,0,0,0,2,0,0,0
2,201800003380,NonLethal,NightNoStreetLight,No,NoIntersection,Normal,Other,4,Saturday,21,...,0,0,0,0,0,0,0,0,0,1
3,201800008060,NonLethal,NightNoStreetLight,No,NoIntersection,Normal,2Vehicles-BehindVehicles-Frontal,8,Friday,22,...,0,0,0,0,0,0,1,0,0,1
4,201800033540,NonLethal,Daylight,Yes,NoIntersection,Overcast,3+Vehicles-Multiple,9,Friday,16,...,0,0,0,0,0,0,3,0,0,1


In [23]:
# 모델링용 데이터 분리
X_train = train.drop(columns='Gravity')
y_train = train['Gravity'].copy()

X_test = test.drop(columns='Gravity')
y_test = test['Gravity'].copy()

# accidents, places 범주형 변수 원-핫 인코딩

In [24]:
# 범수형 변수 객체 생성
categorical_cols = X_train.select_dtypes(include=['object', 'string', 'category']).columns.tolist()

categorical_cols

['Light',
 'InAgglomeration',
 'IntersectionType',
 'Weather',
 'CollisionType',
 'DayOfWeek',
 'RoadType',
 'Circulation',
 'LaneNumber',
 'SpecialLane',
 'Slope',
 'Layout',
 'SurfaceCondition',
 'Infrastructure',
 'Localization',
 'SchoolNear']

In [25]:
# 수치형 변수 객체 생성
numeric_cols = X_train.select_dtypes(include=['int64', 'float64']).columns.tolist()

In [26]:
from sklearn.preprocessing import OneHotEncoder

# OneHot Encoder 생성
encoder = OneHotEncoder(handle_unknown='ignore',sparse_output=False)


# 범주형 변수에 fit
encoder.fit(X_train[categorical_cols]);

In [27]:
# Train / Test 변환

X_train_encoded = encoder.transform(X_train[categorical_cols])
X_test_encoded = encoder.transform(X_test[categorical_cols])

# 컬럼명을 자동 생성
encoded_cols = encoder.get_feature_names_out(categorical_cols)

# DataFrame으로 변환 + 컬럼명 지정
X_train_encoded = pd.DataFrame(X_train_encoded,
                               columns=encoded_cols,
                               index=X_train.index)

X_test_encoded = pd.DataFrame(X_test_encoded,
                              columns=encoded_cols,
                              index=X_test.index)


In [28]:
# 인코딩된 범수형 변수와 수치형 변수 merge
X_train_final = pd.concat([X_train[numeric_cols], X_train_encoded],axis=1)
X_test_final = pd.concat([X_test[numeric_cols], X_test_encoded],axis=1)

# 교차검증

In [29]:
# 데이터 형태 확인
X_train_final.shape, X_test_final.shape

((38256, 194), (9565, 194))

In [30]:
y_train.shape, y_test.shape

((38256,), (9565,))

## Random Forest 교차검증

- 5-fold 교차검증 진행

In [31]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedKFold, cross_validate

rf = RandomForestClassifier(n_estimators=300,
                            random_state=19,
                            n_jobs=-1,
                            class_weight='balanced') #balanced적용

skf = StratifiedKFold(n_splits=5,
                      shuffle=True,
                      random_state=19)

In [32]:
# 평가지표
scoring = {'accuracy': 'accuracy',
           'precision': 'precision',
           'recall': 'recall',
           'f1': 'f1',
           'roc_auc': 'roc_auc'}

# 교차검증 실행
rf_cv = cross_validate(rf,
                       X_train_final,
                       y_train,
                       cv=skf,
                       scoring=scoring,
                       n_jobs=-1)

In [33]:
# 성능 확인
for metric in scoring:
    print(f"{metric}: {rf_cv[f'test_{metric}'].mean():.4f}")

accuracy: 0.9420
precision: nan
recall: nan
f1: nan
roc_auc: 0.8246


In [34]:
# fold 성능
for metric in scoring:
    scores = rf_cv[f'test_{metric}']
    print(
        f"{metric}: "
        f"{scores.mean():.4f} ± {scores.std():.4f}")

accuracy: 0.9420 ± 0.0011
precision: nan ± nan
recall: nan ± nan
f1: nan ± nan
roc_auc: 0.8246 ± 0.0097


## Random Forest 학습

In [35]:
# 전체 데이터 RandomForest fit
rf.fit(X_train_final, y_train);

In [36]:
# Test 데이터 평가
y_pred = rf.predict(X_test_final)

In [37]:
y_pred_proba = rf.predict_proba(X_test_final)[:, 0]

- classes_[0] → Lethal    
- classes_[1] → NonLethal

In [38]:
rf.classes_

array(['Lethal', 'NonLethal'], dtype=object)

In [39]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

print("Accuracy :", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred, pos_label='Lethal'))
print("Recall   :", recall_score(y_test, y_pred, pos_label='Lethal'))
print("F1       :", f1_score(y_test, y_pred, pos_label='Lethal'))

print("\nClassification Report")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, y_pred))

Accuracy : 0.9412441191845269
Precision: 0.3880597014925373
Recall   : 0.09774436090225563
F1       : 0.15615615615615616

Classification Report
              precision    recall  f1-score   support

      Lethal       0.39      0.10      0.16       532
   NonLethal       0.95      0.99      0.97      9033

    accuracy                           0.94      9565
   macro avg       0.67      0.54      0.56      9565
weighted avg       0.92      0.94      0.92      9565


Confusion Matrix
[[  52  480]
 [  82 8951]]


In [40]:
from sklearn.metrics import fbeta_score, precision_score, recall_score

thresholds = [0.1, 0.2, 0.3, 0.4, 0.5]

for threshold in thresholds:
    y_pred_threshold = np.where(
        y_pred_proba >= threshold,
        'Lethal',
        'NonLethal'
    )

    precision = precision_score(
        y_test, y_pred_threshold, pos_label='Lethal'
    )
    
    recall = recall_score(
        y_test, y_pred_threshold, pos_label='Lethal'
    )
    
    f2 = fbeta_score(
        y_test, y_pred_threshold, beta=2, pos_label='Lethal'
    )

    print(
        f"Threshold={threshold:.1f} | "
        f"Precision={precision:.3f} | "
        f"Recall={recall:.3f} | "
        f"F2={f2:.3f}"
    )

Threshold=0.1 | Precision=0.122 | Recall=0.840 | F2=0.386
Threshold=0.2 | Precision=0.186 | Recall=0.618 | F2=0.422
Threshold=0.3 | Precision=0.260 | Recall=0.391 | F2=0.355
Threshold=0.4 | Precision=0.303 | Recall=0.197 | F2=0.212
Threshold=0.5 | Precision=0.388 | Recall=0.098 | F2=0.115


Random Forest는 기본 threshold 0.5에서 Accuracy 94.1%로 높은 성능을 보였지만, Lethal 클래스의 Recall이 9.4%로 매우 낮아 실제 치명적 사고를 탐지하는 데 한계가 있었다. 클래스 불균형의 영향을 고려하여 threshold를 조정한 결과, threshold 0.2에서 Recall 62.8%, F2 0.427로 개선되었으며, 이는 Lethal 탐지 성능과 오탐 간의 균형이 가장 적절한 지점으로 확인되었다.

## XGBoost 교차검증

In [41]:
# target을 0/1로 변환하기
y_train_xgb = y_train.map({'NonLethal' : 0, 'Lethal' : 1})
y_test_xgb = y_test.map({'NonLethal' : 0, 'Lethal' : 1})

In [42]:
# XGBoost 모델 생성
from xgboost import XGBClassifier

scale_pos_weight = ((y_train_xgb == 0).sum() / (y_train_xgb == 1).sum())

scale_pos_weight

np.float64(16.977443609022558)

In [43]:
# 하이퍼파라미터
xgb = XGBClassifier(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=6,
    subsample=0.8,
    colsample_bytree=0.8,
    scale_pos_weight=scale_pos_weight,
    objective='binary:logistic',
    eval_metric='logloss',
    random_state=19,
    n_jobs=-1
)

In [44]:
from sklearn.model_selection import StratifiedKFold, cross_validate
from sklearn.metrics import (
    make_scorer,
    precision_score,
    recall_score,
    f1_score
)

skf = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=19
)

scoring = {
    'accuracy': 'accuracy',
    'precision': make_scorer(
        precision_score,
        pos_label=1
    ),
    'recall': make_scorer(
        recall_score,
        pos_label=1
    ),
    'f1': make_scorer(
        f1_score,
        pos_label=1
    ),
    'roc_auc': 'roc_auc'
}

xgb_cv = cross_validate(
    xgb,
    X_train_final,
    y_train_xgb,
    cv=skf,
    scoring=scoring,
    n_jobs=-1
)

In [45]:
# CV 결과 평균
for metric in scoring.keys():
    scores = xgb_cv[f'test_{metric}']
    
    print(
        f"{metric.upper():10s} : "
        f"{scores.mean():.4f} ± {scores.std():.4f}"
    )

ACCURACY   : 0.8290 ± 0.0017
PRECISION  : 0.1853 ± 0.0054
RECALL     : 0.6109 ± 0.0278
F1         : 0.2843 ± 0.0093
ROC_AUC    : 0.8223 ± 0.0104


In [49]:
# 전체 Train 데이터로 최종 학습
xgb.fit(X_train_final, y_train_xgb);

In [50]:
# Test 데이터 예측
y_pred_xgb = xgb.predict(X_test_final)
y_pred_proba_xgb = xgb.predict_proba(X_test_final)[:, 1]

In [51]:
# 평가지표

print("Accuracy :", accuracy_score(y_test_xgb, y_pred_xgb))
print("Precision:", precision_score(y_test_xgb, y_pred_xgb))
print("Recall:", recall_score(y_test_xgb, y_pred_xgb))
print("F1:",f1_score(y_test_xgb, y_pred_xgb))
print("\nClassification Report")
print(classification_report(y_test_xgb, y_pred_xgb, target_names=['NonLethal', 'Lethal']))
print("\nConfusion Matrix")
print(confusion_matrix(y_test_xgb, y_pred_xgb))

Accuracy : 0.8233141662310507
Precision: 0.18360655737704917
Recall: 0.631578947368421
F1: 0.28450465707027944

Classification Report
              precision    recall  f1-score   support

   NonLethal       0.97      0.83      0.90      9033
      Lethal       0.18      0.63      0.28       532

    accuracy                           0.82      9565
   macro avg       0.58      0.73      0.59      9565
weighted avg       0.93      0.82      0.87      9565


Confusion Matrix
[[7539 1494]
 [ 196  336]]


In [ ]:
# # Threshold 확인
# import numpy as np
# from sklearn.metrics import precision_score, recall_score, fbeta_score

# thresholds = [0.1, 0.2, 0.3, 0.4, 0.5]

# for threshold in thresholds:
    
#     y_pred_threshold = np.where(
#         y_pred_proba_xgb >= threshold,
#         1,
#         0
#     )
    
#     precision = precision_score(
#         y_test_xgb,
#         y_pred_threshold
#     )
    
#     recall = recall_score(
#         y_test_xgb,
#         y_pred_threshold
#     )
    
#     f2 = fbeta_score(
#         y_test_xgb,
#         y_pred_threshold,
#         beta=2
#     )

#     print(
#         f"Threshold={threshold:.1f} | "
#         f"Precision={precision:.3f} | "
#         f"Recall={recall:.3f} | "
#         f"F2={f2:.3f}"
#     )

In [ ]:
from sklearn.metrics import roc_auc_score

roc_auc = roc_auc_score(
    y_test_xgb,
    y_pred_proba_xgb
)

print("ROC-AUC:", roc_auc)